svm/src/runtime_message.rs

```diff
@@ -0,0 +1,68 @@
+use solana_sdk::{message::SanitizedMessage, pubkey::Pubkey};
+
+pub trait RuntimeMessage {
+    /// Return the number of instructions in the message.
+    fn num_instructions(&self) -> usize;
+
+    /// Return an iterator over the instructions in the message.
+    fn instructions_iter(&self) -> impl Iterator<Item = Instruction>;
+
+    /// Return an iterator over the instructions in the message, paired with
+    /// the pubkey of the program.
+    fn program_instructions_iter(&self) -> impl Iterator<Item = (&Pubkey, Instruction)>;
+
+    /// Returns `true` if the account at `index` is writable.
+    fn is_writable(&self, index: usize) -> bool;
+
+    /// Returns `true` if the account at `index` is signer.
+    fn is_signer(&self, index: usize) -> bool;
+}
+
+/// A non-owning version of [`CompiledInstruction`] that references
+/// slices of account indexes and data
+///
+/// [`Message`]: crate::message::Message
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Instruction<'a> {
+    /// Index into the transaction keys array indicating the program account that executes this instruction.
+    pub program_id_index: u8,
+    /// Ordered indices into the transaction keys array indicating which accounts to pass to the program.
+    pub accounts: &'a [u8],
+    /// The program input data.
+    pub data: &'a [u8],
+}
+
+// Implement for the "reference" `SanitizedMessage` type.
+impl RuntimeMessage for SanitizedMessage {
```

Comment:
Not super happy about having an impl for an sdk type here but not sure there's a great way around that.
If we defined the trait somewhere else, let's say its' own crate, it would have to depend on SDK anyway since we will need to return SDK types like `Pubkey`.

I really don't want to add more to SDK, so I think defining it in SVM is not terrible.

Reply:
Cursed. We run into issues when SVM calls into SDK structs like `FeeStructure` which assume `SanitizedMessage`.
Honestly might have to define these traits in SDK given that 🫠